proposals/0118-partitioned-epoch-reward-distribution.md

```diff
@@ -0,0 +1,273 @@
+include: the parameters needed to recalculate the reward partitions, the total
+rewards to be distributed, and the amount of rewards distributed so far.
+
+The layout of `EpochRewards` sysvar is shown in the following pseudo code.
+Prefixes indidate the offset of each field, and padding bytes are always zero.
+
+```
+#[repr(C)] // C representation, 8-byte alignment
+struct EpochRewards{
+   // whether the rewards period (calculation and distribution) is active
+   // byte. false: 0x00, true: 0x01
+   /* 0x00 */ active: bool,
+
+   // little-endian unsigned 64-bit integer
+   /* 0x08 */ distribution_starting_block_height: u64,
+   
+   // little-endian unsigned 64-bit integer
+   /* 0x10 */ num_partitions: u64,
+   
+   // byte-array of length 32
+   /* 0x18 */ parent_blockhash: Hash,
+
+   // total points calculated for the current epoch, where points equals the sum
+   // of delegated stake * credits observed for all delegations
+   // little-endian unsigned 128-bit integer
+   /* 0x38 */ total_points: u128, 
+
+   // total rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x48 */ total_rewards: u64,
+
+   // distributed rewards for the current epoch, in lamports
+   // little-endian unsigned 64-bit integer
+   /* 0x50 */ distributed_rewards: u64, 
+}
+```
+
+The `EpochRewards` sysvar is repopulated at the start of the first block of the
+epoch (before any transactions are processed), as both the total epoch rewards
+and vote account rewards become available at this time. The
+`distributed_rewards` field is updated per reward distribution for each block in
+the reward distribution phase. The `active` field is set to false after the last
+partition is distributed (and `total_rewards == distributed_rewards`), ie.
+before processing transactions in block at height `X + M`.
+
+#### Booting from Snapshot
+
+When booting from a snapshot, a node must check the EpochRewards sysvar account
+to determine whether the distribution phase is active. If so, the node must
+rerun the rewards partitioning using the `EpochRewards::num_partitions` and
+`EpochRewards::parent_blockhash` sysvar fields and determining the upcoming
+partitions by comparing its current block height to
+`EpochRewards::distribution_starting_block_height`. Then the runtime must
+recalculate the remaining rewards using the `EpochRewards::total_points` and
+`EpochRewards::total_rewards` sysvar fields, as well as the `EpochStakes` in the
+snapshot. The recalculated rewards can be confirmed by comparing a sum of the
+rewards remaining (those partitions expected to not yet have been distributed)
+with the difference between the `EpochRewards::total_rewards` and
+`EpochRewards::distributed_rewards` fields. Partitions for blocks prior to the
+current block height can be discarded.
+
+### Restrict Stake Account Mutation
+
+The stake accounts need to be preserved in credit-only state while distributions
```

Comment:
I do not understand the issue here. If the account rewards are credited at the end of the block, then all edits can proceed as normal in txns. Can you elaborate?

Reply:
In this design, if a node boots from snapshot in the middle of the rewards period, it needs to recalculate the reward partitions. Therefore, Delegations need to remain unchanged until all partitions are distributed, as do the number and addresses of stake accounts.

The original implementation of SIMD-0015 prevented any use of the stake program at all during partitioned distribution. However, given the existing use-cases for depositing into stake accounts (MEV), we wanted to relax this restriction. Credit-only is a conservative approach to doing so.

Certainly merges and splits must be prevented during distribution, and we would have to be very careful about allowing delegate or deactivate. But other operations could be permitted.

That said, I lean toward establishing this SIMD with the conservative credit-only approach and amending the SIMD later as we feel confident about the safety of other operations. Wdyt?

Reply:
That is fine for now. This text should make it into the SIMD

Reply:
Pushed

Reply:
That looks good